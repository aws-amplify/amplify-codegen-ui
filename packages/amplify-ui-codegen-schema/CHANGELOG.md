# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2021-10-27)


### Bug Fixes

* adding support for additional component types for string and text types ([53d5537](https://github.com/aws-amplify/amplify-codegen-ui/commit/53d5537f3fd0eca1313d0cd39277ecf297988551))
* moving amplify-ui core dependency to devDependencies to unblock studio build ([3737db2](https://github.com/aws-amplify/amplify-codegen-ui/commit/3737db213c13b948d30197f0638f07fda739fe7a))
* setting license, author, homepage, and repo information ([e253a15](https://github.com/aws-amplify/amplify-codegen-ui/commit/e253a155f36c3451e7bc911225b8757b3dfd8b78))
* update theme generation for new schema ([#142](https://github.com/aws-amplify/amplify-codegen-ui/issues/142)) ([a780893](https://github.com/aws-amplify/amplify-codegen-ui/commit/a7808934e3bb293068687526915a27a3ec8e7637))
* updates to get concat and conditional working, and adding tests ([ef4600f](https://github.com/aws-amplify/amplify-codegen-ui/commit/ef4600f78934b031830f450566b476c2d98caeba))


### Features

* add base action binding support ([#124](https://github.com/aws-amplify/amplify-codegen-ui/issues/124)) ([e6e60c0](https://github.com/aws-amplify/amplify-codegen-ui/commit/e6e60c0394036065991920622bc30caac00dafed))
* concatenation binding implementation ([#99](https://github.com/aws-amplify/amplify-codegen-ui/issues/99)) ([1bfd428](https://github.com/aws-amplify/amplify-codegen-ui/commit/1bfd4287acf7b2d5f410f045e17658929cb60eb3))
* extend base action binding types with navigation types and add test ([dbccfbd](https://github.com/aws-amplify/amplify-codegen-ui/commit/dbccfbd0466186c8cc09d71419504b0ee3abc4ff))
* output theme file ([#97](https://github.com/aws-amplify/amplify-codegen-ui/issues/97)) ([02508c1](https://github.com/aws-amplify/amplify-codegen-ui/commit/02508c1e8733ccee6a17551fed3b885619d70aa7))





# 0.1.0 (2021-10-20)


### Bug Fixes

* moving amplify-ui core dependency to devDependencies to unblock studio build ([3737db2](https://github.com/aws-amplify/amplify-codegen-ui/commit/3737db213c13b948d30197f0638f07fda739fe7a))
* setting license, author, homepage, and repo information ([e253a15](https://github.com/aws-amplify/amplify-codegen-ui/commit/e253a155f36c3451e7bc911225b8757b3dfd8b78))
* updates to get concat and conditional working, and adding tests ([ef4600f](https://github.com/aws-amplify/amplify-codegen-ui/commit/ef4600f78934b031830f450566b476c2d98caeba))


### Features

* add base action binding support ([#124](https://github.com/aws-amplify/amplify-codegen-ui/issues/124)) ([e6e60c0](https://github.com/aws-amplify/amplify-codegen-ui/commit/e6e60c0394036065991920622bc30caac00dafed))
* concatenation binding implementation ([#99](https://github.com/aws-amplify/amplify-codegen-ui/issues/99)) ([1bfd428](https://github.com/aws-amplify/amplify-codegen-ui/commit/1bfd4287acf7b2d5f410f045e17658929cb60eb3))
* extend base action binding types with navigation types and add test ([dbccfbd](https://github.com/aws-amplify/amplify-codegen-ui/commit/dbccfbd0466186c8cc09d71419504b0ee3abc4ff))
* output theme file ([#97](https://github.com/aws-amplify/amplify-codegen-ui/issues/97)) ([02508c1](https://github.com/aws-amplify/amplify-codegen-ui/commit/02508c1e8733ccee6a17551fed3b885619d70aa7))
