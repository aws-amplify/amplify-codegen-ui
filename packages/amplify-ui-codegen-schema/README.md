# `amplify-ui-codegen-schema`

> JSON schema type definitions for Code Generator input JSON files.

## Usage

