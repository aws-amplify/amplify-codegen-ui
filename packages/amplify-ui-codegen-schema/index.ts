export * from "./lib/types/index";
export * from "./lib/types/primitives";
export * from "./lib/types/studio-schema";

/*  legacy
export * from "./lib/types/studio-component-property-type";
export * from "./lib/types/studio-component";
*/
