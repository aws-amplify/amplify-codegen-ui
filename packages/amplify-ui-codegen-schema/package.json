{
  "name": "@amzn/amplify-ui-codegen-schema",
  "version": "0.1.0",
  "description": "Type definitions for schema of JSON, which is input to the code generator",
  "author": "Amazon Web Services",
  "homepage": "https://docs.amplify.aws/",
  "repository": "https://github.com/aws-amplify/amplify-codegen-ui.git",
  "license": "Apache-2.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist/**"
  ],
  "scripts": {
    "test": "echo \"No tests\"",
    "test:update": "echo \"No tests\"",
    "build": "tsc",
    "build:watch": "npm run build -- --watch"
  },
  "devDependencies": {
    "rollup": "^2.47.0",
    "source-map-support": "^0.5.19",
    "typescript": "^4.2.4"
  }
}
