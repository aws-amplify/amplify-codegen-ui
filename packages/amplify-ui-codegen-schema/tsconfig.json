{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "sourceMap": true
  },
  "include": ["./lib", "index.ts"],
  "exclude": ["dist"]
}
