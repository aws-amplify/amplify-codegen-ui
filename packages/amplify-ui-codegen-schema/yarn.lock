# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


buffer-from@^1.0.0:
  version "1.1.2"
  resolved "https://registry.yarnpkg.com/buffer-from/-/buffer-from-1.1.2.tgz#2b146a6fd72e80b4f55d255f35ed59a3a9a41bd5"
  integrity sha512-E+XQCRwSbaaiChtv6k6Dwgc+bx+Bs6vuKJHHl5kox/BaKbhiXzqQOwK4cO22yElGp2OCmjwVhT3HmxgyPGnJfQ==

fsevents@~2.3.2:
  version "2.3.2"
  resolved "https://registry.yarnpkg.com/fsevents/-/fsevents-2.3.2.tgz#8a526f78b8fdf4623b709e0b975c52c24c02fd1a"
  integrity sha512-xiqMQR4xAeHTuB9uWm+fFRcIOgKBMiOBP+eXiyT7jsgVCq1bkVygt00oASowB7EdtpOHaaPgKt812P9ab+DDKA==

rollup@^2.47.0:
  version "2.56.3"
  resolved "https://registry.yarnpkg.com/rollup/-/rollup-2.56.3.tgz#b63edadd9851b0d618a6d0e6af8201955a77aeff"
  integrity sha512-Au92NuznFklgQCUcV96iXlxUbHuB1vQMaH76DHl5M11TotjOHwqk9CwcrT78+Tnv4FN9uTBxq6p4EJoYkpyekg==
  optionalDependencies:
    fsevents "~2.3.2"

source-map-support@^0.5.19:
  version "0.5.19"
  resolved "https://registry.yarnpkg.com/source-map-support/-/source-map-support-0.5.19.tgz#a98b62f86dcaf4f67399648c085291ab9e8fed61"
  integrity sha512-Wonm7zOCIJzBGQdB+thsPar0kYuCIzYvxZwlBa87yi/Mdjv7Tip2cyVbLj5o0cFPN4EVkuTwb3GDDyUx2DGnGw==
  dependencies:
    buffer-from "^1.0.0"
    source-map "^0.6.0"

source-map@^0.6.0:
  version "0.6.1"
  resolved "https://registry.yarnpkg.com/source-map/-/source-map-0.6.1.tgz#74722af32e9614e9c287a8d0bbde48b5e2f1a263"
  integrity sha512-UjgapumWlbMhkBgzT7Ykc5YXUT46F0iKu8SGXq0bcwP5dz/h0Plj6enJqjz1Zbq2l5WaqYnrVbwWOWMyF3F47g==

typescript@^4.2.4:
  version "4.4.2"
  resolved "https://registry.yarnpkg.com/typescript/-/typescript-4.4.2.tgz#6d618640d430e3569a1dfb44f7d7e600ced3ee86"
  integrity sha512-gzP+t5W4hdy4c+68bfcv0t400HVJMMd2+H9B7gae1nQlBzCqvrXX+6GL/b3GAgyTH966pzrZ70/fRjwAtZksSQ==
