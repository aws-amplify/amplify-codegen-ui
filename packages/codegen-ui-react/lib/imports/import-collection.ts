/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { ImportDeclaration, factory } from 'typescript';
import path from 'path';
import { ComponentMetadata, InvalidInputError, reservedWords } from '@aws-amplify/codegen-ui';
import { ImportMapping, ImportValue, ImportSource } from './import-mapping';
import { isPrimitive } from '../primitive';
import { createUniqueName } from '../helpers';
import { ReactRenderConfig } from '../react-render-config';

type ImportCollectionConfig = {
  rendererConfig?: ReactRenderConfig;
};

export class ImportCollection {
  constructor(importConfig?: ImportCollectionConfig) {
    this.importedNames = new Set(reservedWords);
    this.rendererConfig = importConfig?.rendererConfig;
  }

  rendererConfig: ReactRenderConfig | undefined;

  importedNames: Set<string>;

  #collection: Map<string, Set<string>> = new Map();

  importAlias: Map<string, Map<string, string>> = new Map();

  ingestComponentMetadata(componentMetadata: ComponentMetadata) {
    Object.values(componentMetadata?.componentNameToTypeMap || {}).forEach((value) => this.importedNames.add(value));

    // Add form fields so we dont reuse the identifier
    if (componentMetadata?.formMetadata) {
      Object.keys(componentMetadata.formMetadata.fieldConfigs).forEach((value) => this.importedNames.add(value));
    }
  }

  private getOperationsPath(operation: 'mutation' | 'query' | 'subscription' | 'fragment') {
    if (this.rendererConfig?.apiConfiguration?.dataApi === 'GraphQL') {
      switch (operation) {
        case 'mutation':
          return this.rendererConfig.apiConfiguration.mutationsFilePath;
        case 'query':
          return this.rendererConfig.apiConfiguration.queriesFilePath;
        case 'subscription':
          return this.rendererConfig.apiConfiguration.subscriptionsFilePath;
        case 'fragment':
          return this.rendererConfig.apiConfiguration.fragmentsFilePath;
        default:
          throw new InvalidInputError(`Unexpected GraphQL operation encountered: ${operation}`);
      }
    }
    throw new InvalidInputError('Render is not configured to utilize GraphQL operations');
  }

  addMappedImport(...importValue: ImportValue[]) {
    importValue.forEach((value) => {
      const importPackage = ImportMapping[value];
      this.addImport(importPackage, value);
    });
  }

  addGraphqlMutationImport(importName: string) {
    return this.addImport(this.getOperationsPath('mutation'), importName);
  }

  addGraphqlQueryImport(importName: string) {
    return this.addImport(this.getOperationsPath('query'), importName);
  }

  addGraphqlSubscriptionImport(importName: string) {
    return this.addImport(this.getOperationsPath('subscription'), importName);
  }

  addModelImport(importName: string) {
    if (this.rendererConfig?.apiConfiguration?.dataApi === 'GraphQL') {
      return this.addImport(`${this.rendererConfig.apiConfiguration.typesFilePath}`, importName);
    }
    return this.addImport(ImportSource.LOCAL_MODELS, importName);
  }

  addImport(packageName: string, importName: string) {
    if (!this.#collection.has(packageName)) {
      this.#collection.set(packageName, new Set());
    }

    const existingPackage = this.#collection.get(packageName);

    if (!existingPackage?.has(importName)) {
      existingPackage?.add(importName);
      if (packageName !== ImportSource.LOCAL_MODELS) {
        this.importedNames.add(importName);
      }
    }

    if (packageName === ImportSource.LOCAL_MODELS) {
      const existingPackageAlias = this.importAlias.get(packageName);
      const existingAlias = existingPackageAlias?.get(importName);
      if (existingAlias) return existingAlias;
      const modelAlias = createUniqueName(
        importName,
        (input) =>
          // Testing if the name is not unique.
          // Props is for testing the primitive types e.g. "TextProps".
          // And test for a matching primitive name value e.g. "Text"
          this.importedNames.has(input) ||
          (input.endsWith('Props') && isPrimitive(input.replace(/Props/g, ''))) ||
          isPrimitive(input),
      );
      if (existingPackageAlias) {
        existingPackageAlias.set(importName, modelAlias);
      } else {
        const aliasMap = new Map();
        aliasMap.set(importName, modelAlias);
        this.importAlias.set(packageName, aliasMap);
      }
      this.importedNames.add(modelAlias);
      return modelAlias;
    }
    return importName;
  }

  removeImportSource(packageImport: ImportSource) {
    this.#collection.delete(packageImport);
  }

  hasPackage(packageName: string) {
    return this.#collection.has(packageName);
  }

  getMappedAlias(packageName: string, importName: string) {
    return this.importAlias.get(packageName)?.get(importName) || importName;
  }

  getMappedModelAlias(importName: string) {
    if (this.rendererConfig?.apiConfiguration?.dataApi === 'GraphQL') {
      return this.getMappedAlias(this.rendererConfig.apiConfiguration?.typesFilePath || '', importName);
    }
    return this.getMappedAlias(ImportSource.LOCAL_MODELS, importName);
  }

  mergeCollections(otherCollection: ImportCollection) {
    otherCollection.#collection.forEach((value, key) => {
      [...value].forEach((singlePackage) => {
        this.addImport(key, singlePackage);
      });
    });
  }

  buildSampleSnippetImports(topComponentName: string): ImportDeclaration[] {
    return [
      factory.createImportDeclaration(
        undefined,
        undefined,
        factory.createImportClause(
          false,
          undefined,
          factory.createNamedImports([
            factory.createImportSpecifier(undefined, factory.createIdentifier(topComponentName)),
          ]),
        ),
        factory.createStringLiteral('./ui-components'),
      ),
    ];
  }

  getAliasMap(): { model: { [modelName: string]: string } } {
    const modelMap: { [modelName: string]: string } = {};
    const modelSet = this.#collection.get(ImportSource.LOCAL_MODELS);
    const modelAliases = this.importAlias.get(ImportSource.LOCAL_MODELS);
    if (modelSet) {
      [...modelSet].forEach((item) => {
        const alias = modelAliases?.get(item);
        if (alias) modelMap[item] = alias;
      });
    }
    return { model: modelMap };
  }

  buildImportStatements(skipReactImport?: boolean): ImportDeclaration[] {
    const importDeclarations = ([] as ImportDeclaration[])
      .concat(
        skipReactImport
          ? []
          : [
              factory.createImportDeclaration(
                undefined,
                undefined,
                factory.createImportClause(
                  false,
                  undefined,
                  factory.createNamespaceImport(factory.createIdentifier('React')),
                ),
                factory.createStringLiteral('react'),
              ),
            ],
      )
      .concat(
        Array.from(this.#collection)
          .filter(([moduleName]) => moduleName)
          .map(([moduleName, imports]) => {
            const namedImports = [...imports].filter((namedImport) => namedImport !== 'default').sort();
            const aliasMap = this.importAlias.get(moduleName);
            if (aliasMap) {
              const importClause = factory.createImportClause(
                false,
                undefined,
                factory.createNamedImports(
                  [...imports].map((item) => {
                    const alias = aliasMap.get(item);
                    return factory.createImportSpecifier(
                      alias && alias !== item ? factory.createIdentifier(item) : undefined,
                      factory.createIdentifier(alias ?? item),
                    );
                  }),
                ),
              );
              return factory.createImportDeclaration(
                undefined,
                undefined,
                importClause,
                factory.createStringLiteral(moduleName),
              );
            }
            return factory.createImportDeclaration(
              undefined,
              undefined,
              factory.createImportClause(
                false,
                // use module name as default import name
                [...imports].indexOf('default') >= 0 ? factory.createIdentifier(path.basename(moduleName)) : undefined,
                factory.createNamedImports(
                  namedImports.map((item) => {
                    return factory.createImportSpecifier(undefined, factory.createIdentifier(item));
                  }),
                ),
              ),
              factory.createStringLiteral(moduleName),
            );
          }),
      );

    return importDeclarations;
  }
}
