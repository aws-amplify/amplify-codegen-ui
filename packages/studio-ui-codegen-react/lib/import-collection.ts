/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import factory, { createImportDeclaration, ImportDeclaration } from 'typescript';
import path from 'path';

export class ImportCollection {
  #collection: Map<string, Set<string>> = new Map();

  addImport(packageName: string, importName: string) {
    if (!this.#collection.has(packageName)) {
      this.#collection.set(packageName, new Set());
    }

    const existingPackage = this.#collection.get(packageName);

    if (!existingPackage?.has(importName)) {
      existingPackage?.add(importName);
    }
  }

  mergeCollections(otherCollection: ImportCollection) {
    for (const [key, value] of otherCollection.#collection) {
      [...value].forEach((singlePackage) => {
        this.addImport(key, singlePackage);
      });
    }
  }

  buildSampleSnippetImports(topComponentName: string): ImportDeclaration[] {
    const importDeclarations: ImportDeclaration[] = [];
    const sampleStudioPath = './ui-components';
    const namedImports = factory.createNamedImports([
      factory.createImportSpecifier(undefined, factory.createIdentifier(topComponentName)),
    ]);

    importDeclarations.push(
      createImportDeclaration(
        undefined,
        undefined,
        factory.createImportClause(undefined, namedImports),
        factory.createStringLiteral(sampleStudioPath),
      ),
    );
    return importDeclarations;
  }

  buildImportStatements(skipReactImport?: boolean): ImportDeclaration[] {
    const importDeclarations = ([] as ImportDeclaration[])
      .concat(
        skipReactImport
          ? []
          : [
              factory.createImportDeclaration(
                undefined,
                undefined,
                factory.createImportClause(factory.createIdentifier('React'), undefined),
                factory.createStringLiteral('react'),
              ),
            ],
      )
      .concat(
        Array.from(this.#collection).map(([moduleName, imports]) => {
          const namedImports = [...imports].filter((namedImport) => namedImport !== 'default').sort();
          return createImportDeclaration(
            undefined,
            undefined,
            factory.createImportClause(
              // use module name as defualt import name
              [...imports].indexOf('default') >= 0 ? factory.createIdentifier(path.basename(moduleName)) : undefined,
              factory.createNamedImports(
                namedImports.map((item) => {
                  return factory.createImportSpecifier(undefined, factory.createIdentifier(item));
                }),
              ),
            ),
            factory.createStringLiteral(moduleName),
          );
        }),
      );

    return importDeclarations;
  }
}
