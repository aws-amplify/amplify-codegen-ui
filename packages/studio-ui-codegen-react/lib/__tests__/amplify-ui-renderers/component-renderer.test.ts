import { StudioComponent } from '@amzn/amplify-ui-codegen-schema';
import { createPrinter, EmitHint, createSourceFile, ScriptTarget, ScriptKind } from 'typescript';
import { ImportCollection } from '../../import-collection';
import { assertASTMatchesSnapshot } from '../__utils__/snapshot-helpers';

import BadgeRenderer from '../../amplify-ui-renderers/badge';
import ButtonRenderer from '../../amplify-ui-renderers/button';
import BoxRenderer from '../../amplify-ui-renderers/box';
import CardRenderer from '../../amplify-ui-renderers/card';
import DividerRenderer from '../../amplify-ui-renderers/divider';
import FlexRenderer from '../../amplify-ui-renderers/flex';
import ImageRenderer from '../../amplify-ui-renderers/image';
import TextRenderer from '../../amplify-ui-renderers/text';
import renderString from '../../amplify-ui-renderers/string';
import CustomComponentRenderer from '../../amplify-ui-renderers/customComponent';
import CollectionRenderer from '../../amplify-ui-renderers/collection';

function testComponentRenderer(
  Renderer:
    | typeof BadgeRenderer
    | typeof ButtonRenderer
    | typeof BoxRenderer
    | typeof CardRenderer
    | typeof DividerRenderer
    | typeof FlexRenderer
    | typeof ImageRenderer
    | typeof TextRenderer
    | typeof CustomComponentRenderer
    | typeof CollectionRenderer,
  component: StudioComponent,
) {
  const renderChildren = jest.fn();
  const renderedComponent = new Renderer(component, new ImportCollection()).renderElement(renderChildren);

  assertASTMatchesSnapshot(renderedComponent);
}

describe('Component Renderers', () => {
  test('BadgeRenderer', () => {
    const component = {
      componentType: 'Badge',
      name: 'MyBadge',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(BadgeRenderer, component);
  });

  test('ButtonRenderer', () => {
    const component = {
      componentType: 'Button',
      name: 'MyButton',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(ButtonRenderer, component);
  });

  test('BoxRenderer', () => {
    const component = {
      componentType: 'Box',
      name: 'MyBox',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(BoxRenderer, component);
  });

  test('CardRenderer', () => {
    const component = {
      componentType: 'Card',
      name: 'MyCard',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(CardRenderer, component);
  });

  test('DividerRenderer', () => {
    const component = {
      componentType: 'Divider',
      name: 'MyDivider',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(DividerRenderer, component);
  });

  test('FlexRenderer', () => {
    const component = {
      componentType: 'Flex',
      name: 'MyFlex',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(FlexRenderer, component);
  });

  test('ImageRenderer', () => {
    const component = {
      componentType: 'Image',
      name: 'MyImage',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(ImageRenderer, component);
  });

  test('TextRenderer', () => {
    const component = {
      componentType: 'Text',
      name: 'MyText',
      properties: { value: { value: 'test' } },
      bindingProperties: {},
    };
    testComponentRenderer(TextRenderer, component);
  });

  test('CustomComponentRenderer', () => {
    const component = {
      componentType: 'Custom',
      name: 'MyCustom',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(CustomComponentRenderer, component);
  });

  test('CollectionRenderer', () => {
    const component = {
      componentType: 'Collection',
      name: 'MyCollection',
      properties: {},
      bindingProperties: {},
    };
    testComponentRenderer(CollectionRenderer, component);
  });

  describe('StringRenderer', () => {
    test('basic', () => {
      const component = {
        componentType: 'String',
        name: 'MyString',
        properties: {
          value: {
            value: 'test',
          },
        },
        bindingProperties: {},
      };

      const renderedString = renderString(component);
      const file = createSourceFile('test.ts', '', ScriptTarget.ES2015, true, ScriptKind.TS);
      const printer = createPrinter();
      expect(printer.printNode(EmitHint.Unspecified, renderedString, file)).toMatchSnapshot();
    });

    test('missing props', () => {
      const component = {
        componentType: 'String',
        name: 'MyString',
        properties: {},
        bindingProperties: {},
      };
      expect(() => renderString(component)).toThrowErrorMatchingSnapshot();
    });
  });
});
