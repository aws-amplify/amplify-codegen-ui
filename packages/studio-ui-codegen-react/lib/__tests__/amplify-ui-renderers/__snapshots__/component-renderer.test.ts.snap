// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Renderers BadgeRenderer 1`] = `"<Badge {...props} {...getOverrideProps(overrides, \\"Badge\\")}></Badge>"`;

exports[`Component Renderers BoxRenderer 1`] = `"<View {...props} {...getOverrideProps(overrides, \\"View\\")}></View>"`;

exports[`Component Renderers ButtonRenderer 1`] = `"<Button {...props} {...getOverrideProps(overrides, \\"Button\\")}></Button>"`;

exports[`Component Renderers CardRenderer 1`] = `"<Card {...props} {...getOverrideProps(overrides, \\"Card\\")}></Card>"`;

exports[`Component Renderers CollectionRenderer 1`] = `"<Collection items={items || []} {...props} {...getOverrideProps(overrides, \\"Collection\\")}>{(item, index) => ()}</Collection>"`;

exports[`Component Renderers CustomComponentRenderer 1`] = `"<Custom {...findChildOverrides(props.overrides, \\"Custom\\")}></Custom>"`;

exports[`Component Renderers DividerRenderer 1`] = `"<Divider {...props} {...getOverrideProps(overrides, \\"Divider\\")}></Divider>"`;

exports[`Component Renderers FlexRenderer 1`] = `"<Flex {...props} {...getOverrideProps(overrides, \\"Flex\\")}></Flex>"`;

exports[`Component Renderers ImageRenderer 1`] = `"<Image {...props} {...getOverrideProps(overrides, \\"Image\\")}></Image>"`;

exports[`Component Renderers StringRenderer basic 1`] = `"<>{\\"test\\"}</>"`;

exports[`Component Renderers StringRenderer missing props 1`] = `"Failed to render String - Unexpected component structure"`;

exports[`Component Renderers TextRenderer 1`] = `"<Text {...props} {...getOverrideProps(overrides, \\"Text\\")}>{\\"test\\"}</Text>"`;
